java/feathub-connectors/flink-connectors/flink-connector-redis/src/main/java/com/alibaba/feathub/flink/connectors/redis/lookup/RedisLookupFunction.java

```diff
@@ -0,0 +1,191 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.feathub.flink.connectors.redis.lookup;
+
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.flink.table.catalog.ResolvedSchema;
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.functions.FunctionContext;
+import org.apache.flink.table.functions.LookupFunction;
+import org.apache.flink.table.types.CollectionDataType;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.KeyValueDataType;
+import org.apache.flink.util.Preconditions;
+
+import com.alibaba.feathub.flink.connectors.redis.ConversionUtils;
+import com.alibaba.feathub.flink.connectors.redis.JedisClient;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import static com.alibaba.feathub.flink.connectors.redis.RedisConfigs.KEY_FIELDS;
+
+/**
+ * A {@link org.apache.flink.streaming.api.functions.source.SourceFunction} used to query a Redis
+ * database in a lookup join.
+ *
+ * <p>Apart from the columns specified in the configuration keyFields(original key fields), The
+ * input table should contain an even number of columns. half of the columns should have the prefix
+ * "__KEY__" in their column names(redis key fields) and, once the prefix is removed, their columns
+ * names should be equal to those of the other half(redis value fields).
+ *
+ * <p>Each lookup join should pass in the following key fields in order:
+ *
+ * <ul>
+ *   <li>All original key fields, whose order should be the same as that in the keyFields
+ *       configuration value, and
+ *   <li>All redis key fields, whose order should be the same as the order of the redis value fields
+ *       in Redis table's schema.
+ * </ul>
+ *
+ * <p>During a lookup join, the original and redis key fields in the output row would be fulfilled
+ * by directly passing the values from input row. The redis value fields would be fulfilled by
+ * querying against the Redis database, where the keys in the queries comes from the corresponding
+ * redis key fields.
+ */
+public class RedisLookupFunction extends LookupFunction {
+
+    private final ReadableConfig config;
+
+    private final int[] originalKeyFieldIndices;
+    private final int[] redisKeyFieldIndices;
+    private final int[] redisValueFieldIndices;
+
+    private final DataType[] fieldTypes;
+
+    private transient JedisClient client;
+
+    private transient RowData.FieldGetter[] originalKeyGetters;
+
+    private transient RowData.FieldGetter[] redisKeyGetters;
+
+    public RedisLookupFunction(ReadableConfig config, ResolvedSchema schema) {
+        this.config = config;
+        this.originalKeyFieldIndices = getOriginalKeyFieldIndices(schema, config);
```

Comment:
How about using the following names:

logicalKeyFieldIndices
physicalKeyFieldIndices
featureFieldIndices
fieldDataTypes